src/lib/automaton/nfa.mbt

```diff
@@ -9,18 +9,40 @@ typealias Tag = Int
 
 ///|
 pub(all) struct NFA {
-  graph : Array[Map[Symbol, Array[(Int, Tag?)]]] // indexed by node_id * char
+  graph : Array[NFANode]
   end_nodes : Map[Int, Int] // from node_id to pattern_id
   code_blocks : Array[CodeBlock] // indexed by pattern_id
   captures : Array[Array[(String, (Tag, Tag))]] // indexed_by pattern_id
   mut node_count : Int
   mut tag_count : Int
 } derive(Show)
 
+///|
+pub(all) struct NFANode {
+  num : Int
+  mut eps : @immut/sorted_set.T[(NFANode, Tag?)]
+  mut trans : Array[(Char, @immut/sorted_set.T[(NFANode, Tag?)])]
+} derive(Show)
+
+///|
+pub impl Eq for NFANode with op_equal(self, other) {
+  self.num == other.num
+}
+
+///|
+pub impl Compare for NFANode with compare(self, other) {
+  self.num.compare(other.num)
+}
+
+///|
+pub impl Hash for NFANode with hash_combine(self, hasher) {
+  hasher.combine_int(self.num)
+}
+
 ///|
 fn NFA::new_nfa() -> NFA {
   {
-    graph: [Map::new()],
+    graph: [NFANode::{ num: 0, eps: @immut/sorted_set.new(), trans: [] }],
```

Comment:
@LittleJianCH why need this dummy node? I tried remove it, but results unexpected error.